packages/core/src/memo.ts

```diff
@@ -0,0 +1,29 @@
+import type { StateNode } from '.';
+
+const stateNodeMap = new WeakMap<StateNode, any>();
+
+export function getMemo<T>(
+  stateNode: StateNode<any, any>,
+  key: string
+): T | undefined {
+  return stateNodeMap.get(stateNode)?.[key];
+}
+
+export function memo<T>(
+  stateNode: StateNode<any, any>,
+  key: string,
+  fn: () => T
+): T {
+  let memoizedData = stateNodeMap.get(stateNode);
+
+  if (!memoizedData || !memoizedData[key]) {
+    memoizedData = {
+      ...memoizedData,
+      [key]: fn()
+    };
```

Comment:
there is also no reason to use immutability here - let's save some perf here by avoiding spread